docs/src/developer/conventions.md

```diff
@@ -0,0 +1,173 @@
+```@meta
+CurrentModule = Nemo
+```
+
+# Conventions
+
+AbstractAlgebra and Nemo have adopted a number of conventions to help maintain
+a uniform codebase.
+
+## Code conventions
+
+### Function and type names
+
+Names of types in Julia follow the convention of CamelCase where the first
+letter of each word is capitalised, e.g. `Int64` and `AbstractString`.
+
+Function/method names in Julia use all lowercase with underscores between
+the words, e.g. `zip` and `jacobi_symbol`.
+
+We follow these conventions in Nemo with some exceptions:
+
+* When interfacing C libraries the types use the same spelling and
+capitalisation in Nemo as they do in C, e.g. the Flint library's `fmpz_poly`
+remains uncapitalised in Nemo.
+
+* Types such as `gfp_poly` which don't exist under that name on the C side
+also use the lowercase convention as they wrap an actual C type which must be
+split into more than one type on the Julia side. For example `nmod_poly` and
+`gfp_poly` on the Julia side both represent Flint `nmod_poly`'s on the C side.
+
+* Types of rings and fields, modules, maps, etc. are capitalised whether they
+correspond to a C type or not, e.g. `FqNmodFiniteField` for the type of an
+object representing the field that `fq_nmod`'s belong to.
+.
+* We omit an underscore if the first word of a method is "is" or "has", e.g.
+`iseven`.
+
+* Underscores are omitted if the method name is already well established
+without an underscore in Julia itself, e.g. `setindex`.
+
+* Inner and outer constructors with the same name as a type use the same
+spelling and capitalisation as that type, e.g. `fmpz(1)`.
+
```

Comment:
I think I understand what you mean here, although it might appear a bit confusing as by definition, a type constructor (inner or outer) has to have exactly the same name as the type (and there is no first-class distinction between a type and its methods) so it's not like there is any choice. IOW, the type constructor for a type `fmpz` is expected to by `fmpz(...)` for any Julia user; having be `Fmpz` would be totally unexptected. So I think this bullet is superfluous, but YMMV...

Reply:
I'll remove the words inner and outer, but I see no reason why an outer constructor could not be given a different name to the type. This practice is one we want to avoid, so I'm going to leave the point in.

Reply:
There is a precise and a loose meaning for constructor; when qualified with "inner"/"outer", it would typically refer to the precise meaning of a method for a type constructing an instance of that type. The "loose" meaning would be a method of any type or function which creates an instance.